File: tests/integration/Vault/test_set_next_strategy.py
import brownie
import pytest
import time

ZERO_ADDRESS = "0x0000000000000000000000000000000000000000"


def test_current_strategy_is_set(accounts, controller, Vault, erc20, mockStrategy, MockStrategy):
    admin = accounts[0]
    strategy = mockStrategy
    minWaitTime = 1

    vault = Vault.deploy(
        erc20, "vault", "vault", minWaitTime, {'from': accounts[0]}
    )

    # check vault state
    assert vault.timeLock() == 0
    assert vault.minWaitTime() == minWaitTime
    assert vault.strategy() == ZERO_ADDRESS

    # setup
    strategy._setVault_(vault)
    strategy._setUnderlyingToken_(vault.token())

    # set strategy
    vault.setNextStrategy(strategy, {'from': admin})
    time.sleep(minWaitTime)
    vault.switchStrategy({'from': admin})

    assert vault.strategy() == strategy

    nextStrategy = MockStrategy.deploy(
        controller, vault, vault.token(), {'from': admin}
    )

    def get_snapshot():
        return {
            "vault": {
                "strategy": vault.strategy(),
                "nextStrategy": vault.nextStrategy(),
                "timeLock": vault.timeLock(),
            }
        }

    before = get_snapshot()
    tx = vault.setNextStrategy(nextStrategy, {'from': admin})
    after = get_snapshot()

    assert after["vault"]["nextStrategy"] == nextStrategy
    assert after["vault"]["timeLock"] == tx.timestamp + minWaitTime

File: test/unit/Vault/test-set-next-strategy.ts
import chai from "chai"
import BN from "bn.js"
import {Erc20TokenInstance} from "../../../types/Erc20Token"
import {VaultInstance} from "../../../types/Vault"
import {MockControllerInstance} from "../../../types/MockController"
import {StrategyTestInstance} from "../../../types/StrategyTest"
import {ZERO_ADDRESS, eq, getBlockTimestamp} from "../../util"
import _setup from "./setup"

const StrategyTest = artifacts.require("StrategyTest")

contract("Vault", (accounts) => {
  const MIN_WAIT_TIME = 100

  const refs = _setup(accounts, MIN_WAIT_TIME)
  const {admin} = refs

  let controller: MockControllerInstance
  let vault: VaultInstance
  let erc20: Erc20TokenInstance
  let strategy: StrategyTestInstance
  beforeEach(() => {
    controller = refs.controller
    vault = refs.vault
    erc20 = refs.erc20
    strategy = refs.strategy
  })

  describe("setNextStrategy", () => {
    it("should set next strategy when current strategy is not set", async () => {
      const tx = await vault.setNextStrategy(strategy.address, {from: admin})

      assert.equal(tx.logs[0].event, "SetNextStrategy", "event")
      assert.equal(
        // @ts-ignore
        tx.logs[0].args.strategy,
        strategy.address,
        "event arg next strategy"
      )
      assert.equal(await vault.nextStrategy(), strategy.address)
      assert(eq(await vault.timeLock(), new BN(0)), "time lock")
    })

    it("should set next strategy when current strategy is set", async () => {
      await vault.setNextStrategy(strategy.address, {from: admin})
      await vault.setStrategy(strategy.address, {from: admin})

      assert.equal(await vault.strategy(), strategy.address, "strategy")

      const newStrategy = await StrategyTest.new(
        controller.address,
        vault.address,
        erc20.address,
        {from: admin}
      )
      const tx = await vault.setNextStrategy(newStrategy.address, {
        from: admin,
      })

      const timestamp = await getBlockTimestamp(web3, tx)

      assert.equal(await vault.nextStrategy(), newStrategy.address, "next strategy")
      assert(eq(await vault.timeLock(), new BN(timestamp + MIN_WAIT_TIME)), "time lock")
    })

    it("should reject if not admin", async () => {
      await chai
        .expect(vault.setNextStrategy(strategy.address, {from: accounts[1]}))
        .to.be.rejectedWith("!admin")
    })

    it("should reject zero address", async () => {
      await chai
        .expect(vault.setNextStrategy(ZERO_ADDRESS, {from: accounts[1]}))
        .to.be.rejectedWith("!admin")
    })

    it("should reject same strategy", async () => {
      await vault.setNextStrategy(strategy.address, {from: admin})

      await chai
        .expect(vault.setNextStrategy(strategy.address, {from: admin}))
        .to.be.rejectedWith("same next strategy")
    })
  })
})
